[Concept] Docs - References (#663)
exercism/cpp: concepts/references/introduction.md

# Introduction

## Copies and References

With our current knowledge, it is a bit cumbersome to update a variable with the help of a function.
Every argument is handled like a copy and does not change the original.
Depending on the size of your type, this can have serious consequences for the performance of your code.

```cpp
int adjust_salary(int base, int kids) {
    return current + kids * 500;
}
int now_a_mother_of_twins{2500};
adjust_salary(now_a_mother_of_twins, 2);

// now_a_mother_of_twins is still 2500
now_a_mother_of_twins = adjust_salary(now_a_mother_of_twins, 2);
// now_a_mother_of_twins is now 3500
```

`References` can be seen like aliases, changes to them have an effect on the original variable.
`References` use the **ampersand** in the type declaration: `&`.

```cpp
int balance{1000};
int& budget{balance};
int pro_computer_wheels{699};
budget -= pro_computer_wheels;
// budget is now 301;
// balance is also 301;
```

> Reseating - changing the binding of a `reference` is not possible.

Reviewer: Found this SO post interesting wrt this, but might be too much to put in the docs at this stage: https://stackoverflow.com/questions/728233/why-are-references-not-reseatable-in-c

> Reseating - changing the binding of a `reference` is not possible.

Author: I might use it for the about section as a "read more" piece.

> You cannot have an unitialized `reference`, you always need a binding.

Reviewer: ```suggestion
You cannot have an uninitialized `reference`, you must always have a binding for it.
```

> You cannot have an unitialized `reference`, you always need a binding.

Reviewer: IMHO, `initialize` make things more confusing - maybe
 "it always needs to be connected to a variable"?  
Or perhaps
 "it always needs to reference (connect to) a variable"?

> You cannot have an unitialized `reference`, you always need a binding.

Reviewer 2: "initialization" is a technical term in C++ (and is a surprisingly complex topic, one of the top experts wrote a whole [book](https://www.cppstories.com/2022/cpp-init-book/) about it).  
it was mentioned briefly in the "Basics" concept.

"connect" on the other hand is not a well-defined term in C++ (I think)

> You cannot have an unitialized `reference`, you always need a binding.

Reviewer: I think for beginners, "initialize" often means "starter value".  But once initialized, `references` can't be changed - or at least that is what I am grokking from the text.  Hence my reticence in using it.  Binding feels less confusing to me, since "bind" in English means "to tie together".

> You cannot have an unitialized `reference`, you always need a binding.

Reviewer 2: Yes, "starter value" sounds about right (colloquially speaking). It's just that you have to provide that starter value (you can't leave references uninitialized), and you cannot re-assign the reference so that it refers to a different object.  
The term "binding" is not wrong, we just haven't introduced it yet.

> You cannot have an unitialized `reference`, you always need a binding.

Author: I always had it as binding in my head, but some research suggests, that it is not as official as I thought it to be.
Students will be used to initializations by now. I would not go with "starter value" because that sounds like it is a temporary thing that might change.

```cpp
int main_acc{1'000'000};
int side_acc{-20};
int& savings{main_acc};
// => main_acc and savings are 1'000'000

// try to reseat savings to use main account
savings = side_acc;
// savings and main_acc are now -20
// as this uses the **value** of side_acc
savings += 20;
// savings and main_acc are now 0, side_acc is stil -20

int& future_budget;
// => compiler error, reference must be bound!
```

## `void`

> With the power of `references` you might not have the need to return a value from a function. This is a scenario to use `void` as a return type.

Reviewer: ```suggestion
With the power of `references` you might not need to return a value from a function at all. 
`void` is used as a return type in this scenario. 
```
